src/main/java/org/elasticsearch/cluster/action/index/NodeIndexDeletedAction.java

```diff
@@ -64,32 +75,56 @@ public void remove(Listener listener) {
     }
 
     public void nodeIndexDeleted(final ClusterState clusterState, final String index, final String nodeId) throws ElasticsearchException {
-        DiscoveryNodes nodes = clusterState.nodes();
+        final DiscoveryNodes nodes = clusterState.nodes();
         if (nodes.localNodeMaster()) {
-            threadPool.generic().execute(new Runnable() {
+            threadPool.generic().execute(new AbstractRunnable() {
+
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warn("failed to ack index store deleted for  index {}", t, index);
+                }
+
                 @Override
-                public void run() {
+                protected void doRun() throws Exception {
                     innerNodeIndexDeleted(index, nodeId);
+                    lockIndexAndAck(index, nodes, nodeId, clusterState);
+
                 }
             });
         } else {
             transportService.sendRequest(clusterState.nodes().masterNode(),
                     INDEX_DELETED_ACTION_NAME, new NodeIndexDeletedMessage(index, nodeId), EmptyTransportResponseHandler.INSTANCE_SAME);
+            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
The other side of this action expects store acks from all the data nodes (see https://github.com/elasticsearch/elasticsearch/blob/master/src/main/java/org/elasticsearch/cluster/metadata/MetaDataDeleteIndexService.java#L139 ) but as far as I can tell we call this on every node now. I did some history digging and it seems this has been the case in 1.3 as well. We can fix this and keep the two actions or just use a single action call back which signals "I'm all done" (i.e., index removed from memory and from disk, if needed).